File: 02_stack/00_valid_parentheses/valid_parentheses.cpp
#include <string>
#include <stack>

using std::string;
using std::stack;

class Solution
{
  public:
    bool isValid(string s)
    {
        stack<char> check;
        for (const char& c : s)
        {
            if (c == '(' || c == '[' || c == '{')
                check.push(c);
            else if (c == ')' || c == ']' || c == '}')
            {
                if (check.size() == 0)
                    return false;
                if (check.top() == '(' && c == ')' ||
                    check.top() == '[' && c == ']' ||
                    check.top() == '{' && c == '}')
                    check.pop();
                else
                    return false;
            }
        }
        return check.size() == 0;
    }
};

File: NeetCode/02_stack/00_valid_parentheses/valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        other = {
            ')': '(',
            ']': '[',
            '}': '{'
        }
        check = []
        for c in s:
            if c in "([{":
                check.insert(0, c)
            elif c in other:
                if len(check) == 0 or check[0] != other[c]:
                    return False
                else:
                    check.pop(0)
        return len(check) == 0


if __name__ == "__main__":
    solution = Solution()
    print(solution.isValid("]"))
        